File: leetcode/1663-smallest-string-with-a-given-numeric-value/solution.py
class Solution:
    def getSmallestString(self, n: int, k: int) -> str:
        # every position must hold at least 'a' (value 1), so start with a baseline
        # of n and distribute the remaining value, greedily pushed as far right
        # as possible so the earlier characters stay as small as possible
        result = ["a"] * n
        remaining = k - n

        for i in range(n - 1, -1, -1):
            if remaining <= 0:
                break
            add = min(25, remaining)  # 'a' + add, capped so it never exceeds 'z' (value 26)
            result[i] = chr(ord("a") + add)
            remaining -= add

        return "".join(result)

File: leetcode/1663-smallest-string-with-a-given-numeric-value/solution.cpp
#include <string>
#include <algorithm>
using namespace std;

class Solution {
public:
    string getSmallestString(int n, int k) {
        string result(n, 'a');
        int remaining = k - n;

        for (int i = n - 1; i >= 0 && remaining > 0; i--) {
            int add = min(25, remaining);
            result[i] = (char)('a' + add);
            remaining -= add;
        }

        return result;
    }
};
